Ex 2.1 Mathematical modelling of given assignemnt problem

Let $x_{ij}$ be a decision variable which is used to indicate assignment of factory 'i' at location 'j'.  

This is a binary variable which can take value either 0 or 1.   

$x_{ij}$=0; signifies that factory 'i' is not assigned at location 'j'.  
$x_{ij}$=1; signifies that factory 'i' is assigned at location 'j'.  

The parameter 'i' represents factory no. As there are 12 factories so it can take values {1,2,3....,12}.

The parameter 'j' represents location no. As there are 12 locations so it can take values {1,2,3....,12}.

The cost for assignments of factory 'i' at location 'j' is represented by $c_{ij}$. The cost matrix for each of the assigment is read from file "lab5_ex2.txt" and this value is stored in cost_matrix during the programming.

The constraints are that at any particular location only one factory can be assigned and a factory can be assigned at only one location.   
Constraints:

\begin{align}
\sum_{i=1}^{12} x_{i1} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i2} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i3} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i4} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i5} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i6} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i7} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i8} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i9} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i10} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i11} &= 1  \nonumber \\
\sum_{i=1}^{12} x_{i12} &= 1  \nonumber 
\end{align}

&

\begin{align}
\sum_{j=1}^{12} x_{1j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{2j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{3j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{4j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{5j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{6j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{7j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{8j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{9j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{10j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{11j} &= 1  \nonumber \\
\sum_{j=1}^{12} x_{12j} &= 1  \nonumber 
\end{align}

So we have a total of 24 constraints.

Objective is to minimise the total cost of assignment.
\begin{align}
\min  \sum_{i=1}^{12} \sum_{j=1}^{12} x_{ij}*c_{ij} \nonumber 
\end{align}

So this completes our modelling of the problem.

In [39]:
!pip install -q pyomo
from pyomo.environ import * 
import numpy as np
import pandas as pd
!apt-get install -y -qq coinor-cbc

In [40]:
cost_matrix=np.loadtxt('lab5_ex2.txt', delimiter=',')

N=cost_matrix.shape[1]
M=cost_matrix.shape[0]

col_indices=range(N)
row_indices=range(M)


In [41]:

model=ConcreteModel()


#variable declaration
model.x=Var(col_indices,row_indices,domain=Binary)

#adding constraints
model.constraints=ConstraintList()

for i in row_indices:
  model.constraints.add(expr=sum(model.x[i,j] for j in col_indices)==1)

for j in col_indices:
  model.constraints.add(expr=sum(model.x[i,j] for i in row_indices)==1)

#objective function
model.objective=Objective(expr=sum(model.x[i,j]*cost_matrix[i,j] for i in row_indices for j in col_indices),sense=minimize)

model.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 1

In [42]:
opt_cbc=SolverFactory('cbc')
result=opt_cbc.solve(model)

print('\nObjective = ', model.objective())

print("assigned factory i to  location j")
for i in row_indices:
  for j in col_indices:
    if model.x[i,j]:  
      print("i=",i+1,"j=",j+1)


Objective =  198.0
assigned factory i to  location j
i= 1 j= 2
i= 2 j= 7
i= 3 j= 12
i= 4 j= 10
i= 5 j= 3
i= 6 j= 11
i= 7 j= 4
i= 8 j= 6
i= 9 j= 5
i= 10 j= 1
i= 11 j= 8
i= 12 j= 9


In [43]:
#2.9 removing integer restriction from decision variables
model.x.domain=NonNegativeReals

lower_bound = 0
upper_bound = 1

model.x.setlb(lower_bound)
model.x.setub(upper_bound)

result=opt_cbc.solve(model)

print('\nObjective = ', model.objective())

for i in row_indices:
  for j in col_indices:
    if model.x[i,j]:  
      print("value of decision variable x[",i+1,"][",j+1,"]=",model.x[i,j].value)




Objective =  198.0
value of decision variable x[ 1 ][ 2 ]= 1.0
value of decision variable x[ 2 ][ 7 ]= 1.0
value of decision variable x[ 3 ][ 12 ]= 1.0
value of decision variable x[ 4 ][ 10 ]= 1.0
value of decision variable x[ 5 ][ 3 ]= 1.0
value of decision variable x[ 6 ][ 11 ]= 1.0
value of decision variable x[ 7 ][ 4 ]= 1.0
value of decision variable x[ 8 ][ 6 ]= 1.0
value of decision variable x[ 9 ][ 5 ]= 1.0
value of decision variable x[ 10 ][ 1 ]= 1.0
value of decision variable x[ 11 ][ 8 ]= 1.0
value of decision variable x[ 12 ][ 9 ]= 1.0


2.10 After removing the integer restrictions from decision variables, the optimal cost remains unchanged at 198.   
Yes the value of the decision variables are still integer valued.

The decision variables are still taking integer values mainly because of the constraints and bounds which force the sum of values across any row or column to be 1 and any decision variable can take values between 0 and 1 only.

2.12  facility 1 cannot be assigned to location 4, facility
11 cannot be assigned to location 3 and facility 5 cannot be assigned to location 9

To take this new condition in account, I have assigned very high value(99999) to the cost corresponding to these links. This new cost is considerably large compared to cost for other links, by doing this it is ensured that if any value is assigned to the decision variable for this link then value of objective will increase drastically and hence no value will be assigned to this particular link.

New csv file for this change is "lab5_ex2_new.csv".

Further without changing the model itself we just need to read data from this new file. Old objective function has been deactivated and new objective is introduced.

In [44]:
#introducing new restrictions by using new txt file 
cost_matrix=np.loadtxt('lab5_ex2_new.txt', delimiter=',')

model.x.domain=Binary

model.objective_new=Objective(expr=sum(model.x[i,j]*cost_matrix[i,j] for i in row_indices for j in col_indices),sense=minimize)
model.objective.deactivate()

result=opt_cbc.solve(model)

print('\nObjective = ', model.objective_new())

print("assigned factory i to  location j")
for i in row_indices:
  for j in col_indices:
    if model.x[i,j]:  
      print("i=",i+1,"j=",j+1)


Objective =  198.0
assigned factory i to  location j
i= 1 j= 2
i= 2 j= 7
i= 3 j= 12
i= 4 j= 10
i= 5 j= 3
i= 6 j= 11
i= 7 j= 4
i= 8 j= 6
i= 9 j= 5
i= 10 j= 1
i= 11 j= 8
i= 12 j= 9
